src/emm/services/modules_service.py

```diff
@@ -225,9 +230,113 @@ def git_commit_modules(self, commit: str) -> None:
+        pr_dict: Dict[str, List[str]] = {}
+        for module, pr_url in comments.items():
+            pr_links = "\n".join(
+                "{} pr: {}".format(repo, link) for repo, link in comments.items() if repo != module
+            )
+            pr_dict[module] = []
+            pr_dict[module].append(pr_url)
+            pr_dict[module].append(pr_links)
+        return pr_dict
+
+    def update_pr_links(self, comment_links: Dict[str, List[str]]) -> List[str]:
+        """
+        Update link to each module.
+
+        :param comment_links: Dictionary of module with its location and other modules' pull request link.
+        :return: List of pull requests being created associate with its link.
+        """
+        all_pull_requests = []
+        lines = (
+            "This code review is spread across multiple repositories. "
+            "Links to the other components of this review can be found:\n "
+        )
+        for module, pr_list in comment_links.items():
+            pr_url = pr_list[0]
+            if module != "base":
+                module_data = self.get_module_data(module)
+                module_location = Path(module_data.prefix) / module
+                pr_links = lines + pr_list[1]
+                self.github_service.pr_comment(pr_url, pr_links, module_location)
+                all_pull_requests.append(pr_url)
+                all_pull_requests.append(pr_links)
+            else:
+                pr_links = lines + pr_list[1]
+                self.github_service.pr_comment(pr_url, pr_links)
+                all_pull_requests.append(pr_url)
+                all_pull_requests.append(pr_links)
+
+        return all_pull_requests
+
+    def push_branch_to_remote(self, directory: Optional[Path] = None) -> None:
+        """
+        Determine if need push current branch to remote.
+
+        :param directory: Directory to execute the command at.
+        """
+        branch = self.git_service.current_branch(directory)
+        if branch == "main" or branch == "master":
+            raise ValueError("Cannot push unreviewed modification to master.")
+
+        if not self.git_service.current_branch_exist_on_remote(branch, directory):
+            self.git_service.push_branch_to_remote(directory)
+
+    def module_pull_request(self, args: List[str]) -> Dict[str, str]:
+        """
+        Create pull request for each module.
+
+        :param args: Arguments to pass to the github CLi.
+        :return: Dictionary of module and its pull request url.
+        """
+        comment_dict = {}
+        enabled_modules = self.get_all_modules(True)
+        for module in enabled_modules:
+            module_data = self.get_module_data(module)
+            module_location = Path(module_data.prefix) / module
+            basename = self.git_service.get_base_name(module_location)
+            if basename and self.git_service.check_changes(basename, module_location):
+                self.push_branch_to_remote(module_location)
```

Comment:
The ordering here seems dangerous. This function could raise an exception which would abort the operation. But since we are calling it in the middle of operating on modules, there is a chance that we create PR on some modules then hit an error, leaving things in an inconsistent state. Some of the PR created and others not.

It seems like it would be better to validate that all the modules are in a good state as best we can, before we start creating PRs.

Reply:
Thanks for catching it! I would iterate the enabled_module twice to ensure every branch is created to remote before create a pull request. I assume there won't be too many enabled_modules so loop twice won't impact the overall performance.